### ISSUE TITLE
OutOfMemoryError spins library out of control

### ISSUE THREAD

**Issue by TWiStErRob** [classgraph/classgraph#400]
## Issue summary
`nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler#close` may never complete if Thread was interrupted, which is always if scanning failed. See below for details.

## Background
I was writing some unit tests for a really basic Neo4J graph, but while Neo4J-OGM was looking for my Entities during test setup, it just never completed. With some very roundabout debugging (all the async and OOM interplay was making me tear my hair out), I managed to narrow down the root cause to how classgraph works. I have 32G memory, and Java defaults to using 1/4th of memory per process, which means I cannot run much in parallel, so I have a global 256M override for each process. It works fine for Gradle, Neo4J, and other traditionally resource-hungry setups. It even works for my test setup most of the time, so for this report repro I lowered the memory limit to be consistent.

## Repro Setup
trivial library usage:
```java
	// implementation "io.github.classgraph:classgraph:4.8.62"
	ScanResult result = new ClassGraph()
			.whitelistPackages("pack.age")
			.verbose()
			.scan()
			;
```
plus a dependency with a nested jar in it:
```gradle
runtimeOnly "org.neo4j:neo4j-lucene-upgrade:3.4.9"
```
non-trivial: resource constrained memory: `-Xmx128M` (see build.gradle)

Full minimal repro can be found here: https://github.com/TWiStErRob/repros/tree/master/classgraph/oom-NestedJarHandler-spin

## Repro steps
1. `gradlew run`

or

1. Import project to IntellIJ
1. Execute `Main.main` (should pass, depending on machine setup)
1. Edit Run Configuration to have VM options: `-Xmx128M`
1. Run or debug again (will fail, expected: print a line)

[gradlew run.log](https://github.com/classgraph/classgraph/files/4174299/run.log)

## Pieces of the puzzle
What I found to be contributing to the issue.

### Nested Jar files
```gradle
runtimeOnly "org.neo4j:neo4j-lucene-upgrade:3.4.9"
```
an example from Neo4J/OGM, which contains !/lib/lucene-backward-codecs-5.5.0.jar
This is required to trigger
```java
                    } else {
                        // This path has one or more '!' sections.
```
code path in `NestedJarHandler`.

### Large allocation to trigger OOM
This is baked into the library. `MappedByteBufferResources` allocates 64M per JAR file, which is quite large, considering there could be many jars like this on the classpath, there could a lot of allocations happening at the same time and consuming tons of memory. Note: I have 14 core/28 logical core processor, and the library runs on 39 threads. So if all those are processing nested JARs, 2.5G of memory is required. That doesn't sound "ultra-lightweight" :)

This allocation throws an OOM, which is caught by `AutoCloseableExecutorService.afterExecute` and the thread is interrupted. I think this interrupt doesn't hurt `NestedJarHandler.close`, so it could be an irrelevant red herring to the issue. But this location came up during the investigation so I thought I would mention it.

### `Scanner.call` catches and cleans up
When an exception occurred (e.g. our OOM), `removeTemporaryFilesAfterScan` is set to `true` and threads are interrupted. `finally` then comes and tries to close `nestedJarHandler`. The relevant parts of the method:
```java
    @Override
    public ScanResult call() throws InterruptedException, CancellationException, ExecutionException {
        try {
            scanResult = openClasspathElementsThenScan();
        } catch (final Throwable e) {
            // Since an exception was thrown, remove temporary files
            removeTemporaryFilesAfterScan = true;

            // Stop any running threads (should not be needed, threads should already be quiescent)
            interruptionChecker.interrupt();
        } finally {
            if (removeTemporaryFilesAfterScan) {
                // If removeTemporaryFilesAfterScan was set, remove temp files and close resources,
                // zipfiles and modules
                nestedJarHandler.close(topLevelLog);
            }
        }
        return scanResult;
    }
```

### `NestedJarHandler#close` spins
See numbered inline comments for explanation.
```java
            if (canonicalFileToPhysicalZipFileMap != null) {
                // (4) spins out of control, because (3)
                while (!canonicalFileToPhysicalZipFileMap.isEmpty()) {
                    try {
                        // (1) throws InterruptedException if thread was interrupted, which is always because Scanner error handling interrupted
                        for (final Entry<File, PhysicalZipFile> ent : canonicalFileToPhysicalZipFileMap.entries()) {
                            final PhysicalZipFile physicalZipFile = ent.getValue();
                            physicalZipFile.close();
                            // (3) never executed, because of ConcurrentHashMap's interruption in SingletonMap
                            canonicalFileToPhysicalZipFileMap.remove(ent.getKey());
                        }
                    } catch (final InterruptedException e) {
                        // (2) re-interrupts Thread
                        interruptionChecker.interrupt();
                    }
                }
                canonicalFileToPhysicalZipFileMap = null;
            }
```

## Proposed solutions
1. Quick treatment with library gracefully executing to successful completion:  
    Catch OOM explicitly when allocating `buf` in `MappedByteBufferResources` and spill to disk if that happened. This would be in line with how `IOException` from `makeTempFile` is caught.
2. Lower memory usage (could be combined with previous):  
    `fastZipEntryToZipFileSliceMap` knows the size of the nested Jar file (childZipEntry), if that was passed down the decision can be immediately made to spill to disk before allocating and without reading the whole stream.
3. Don't spin  
   In any case, `NestedJarHandler.close` should probably be able to close `canonicalFileToPhysicalZipFileMap` even in an interrupted state, so that fail classpath scans can receive the exceptional termination and handle accordingly.

**Comment by lukehutch** [comment]
@TWiStErRob you sir deserve a medal for this bug report. I wish all bug reports were this detailed and well-researched!!

> So if all those are processing nested JARs, 2.5G of memory is required. That doesn't sound "ultra-lightweight" :)

Not quite -- firstly this is virtual memory space that is used, not allocated RAM. The full 2.5GB will only become "resident" (swapped in) if the entire range is in constant active use -- and it will be swapped back out to disk if it has changed (or simply dropped from the page table if it hasn't changed and hasn't been used recently, under memory pressure). Also unless you are running multiple copies of the same program in completely different processes, then memory-mapped jars for a single `ClassGraph` instance are shared between threads in the same process (e.g. the ClassGraph worker threads) -- and even if you do have multiple processes mapping the same file to RAM, any modern OS (at least Linux, but I assume also Mac OS X and Windows?) should see that the same file is already mapped when a request to map it again is issued, and re-use the mapping without allocating more virtual memory range. This should also cause mapping reuse across `ClassGraph` instances within or across processes.

It's still possible to run out of virtual memory space easily on a 32-bit OS though, and it's possible to hit an OOM condition by running out of Java memory map file handles on Linux (there's a limit of 64k of them).

That said, ClassGraph can still consume an enormous amount of RAM, simply to build the graph of `ClassInfo` objects, and if you have multiple instances of this, and/or you're scanning a large number of classes, this can quickly consume all RAM (#338, #371). `ClassInfo` objects cannot be reused across `ClassGraph` instances or across processes. I spent a lot of time carefully looking at RAM usage for #338, and reduced RAM usage as much as possible, but there was really no way to reduce it further without the caller taking actions to reduce the size of the `ScanResult` by using more aggressive whitelisting etc. -- so further lowering memory usage in general is not really a possibility.

> This is baked into the library. `MappedByteBufferResources` allocates 64M per JAR file, which is quite large, considering there could be many jars like this on the classpath

This is in tension with the fact that some users of ClassGraph really, really don't want ClassGraph ever writing anything to disk, or perhaps don't even have a writeable temporary directory at all (#236). The 64MB default was intentionally set high to cause ClassGraph to avoid writing to disk wherever possible. Also, what you're suggesting, of lowering the 64M and spilling to disk more readily, would absolutely kill performance for some runtimes with lots of nested jars, e.g. a Spring-Boot application, because right now as long as a nested jar is stored and not deflated, the nested jar can be read directly, using file slicing, when its own entries need to be deflated. If the 64M were lowered, then it would be more common that the inner jar would need to be copied to a temporary file on disk just to have its entries read, which would be wasteful.

Probably though this should be configurable on a per-application basis, so it's a good idea to add a configuration option for it. I'll do that.

On your suggestions:

1. Handling OOM by automatically spilling to disk is a good idea. I'll look into how to do this safely. This will still bring the system close to OOM, but it may be enough to stop the program bombing in many situations.
2. Good suggestion to pass the filesize down.
3. Oops, good catch and great analysis on the spinning issue. This is a pretty serious bug that I'll get working on too.

Lots to do here -- thanks for bringing these issues to my attention!

**Comment by lukehutch** [comment]
Actually I realized the 64MB in-memory default is probably not something you're running into -- see this comment from the latest checkin:

```java
    /**
     * The maximum size of an inner (nested) jar that has been deflated (i.e. compressed, not stored) within an
     * outer jar, before it has to be spilled to disk rather than stored in a RAM-backed {@link ByteBuffer} when it
     * is deflated, in order for the inner jar's entries to be read. (Note that this situation of having to deflate
     * a nested jar to RAM or disk in order to read it is rare, because normally adding a jarfile to another jarfile
     * will store the inner jar, rather than deflate it, because deflating a jarfile does not usually produce any
     * further compression gains. If an inner jar is stored, not deflated, then its zip entries can be read directly
     * using ClassGraph's own zipfile central directory parser, which can use file slicing to extract entries
     * directly from stored nested jars.)
     * 
     * <p>
     * This is also the maximum size of a jar downloaded from an {@code http://} or {@code https://} classpath
     * {@link URL} to RAM. Once this many bytes have been read from the {@link URL}'s {@link InputStream}, then the
     * RAM contents are spilled over to a temporary file on disk, and the rest of the content is downloaded to the
     * temporary file. (This is also rare, because normally there are no {@code http://} or {@code https://}
     * classpath entries.)
     * 
     * <p>
     * Default: 64MB (i.e. writing to disk is avoided wherever possible). Setting a lower max RAM size value will
     * decrease ClassGraph's memory usage if either of the above rare situations occurs.
     */
```

However if you were running into this case, were the inner jars in question deflated, not stored? I want to make sure ClassGraph isn't treating your inner jars as deflated (needing to be extracted) when they were stored (not needing to be extracted).

More likely you are running into a limit on the amount of mapped virtual memory, not the amount of resident memory.

The first of the above two commits implements your suggestion "2.", and also trims the 64MB buffer if the extracted or downloaded jar size is smaller than 64MB (this was an omission).

The second of the above commits allows the 64MB threshold to be configured.

**Comment by lukehutch** [comment]
One more TODO item -- after the worker threads have finished, the worker-specific resources (e.g. file slice mappings) should be freed. This will reduce the memory pressure between when the scan finishes and the `ScanResult` is closed.

**Comment by lukehutch** [comment]
I just checked in a large change that allows files to be backed with `RandomAccessFile` rather than `MappedByteBuffer`, eliminating the virtual memory pressure altogether. I thought there would be a performance penalty for this, but that doesn't seem to be the case, at least on Linux (I guess the filesystem caching works really well) -- so maybe I should remove the memory mapping altogether.

These changes are triggering another big refactoring too, as I see some other ways to simplify and unify some of the code.

**Comment by TWiStErRob** [comment]
Wow, lot of info :) And also I'm not used to getting fixes (let alone replies) this fast. Well done!

Here are a few thoughts while reading them comments/commits.

---
> I hadn't considered though that Java limits not just the resident size but also the entire accessible virtual memory size

Yes, Java OOM means there's no more JVM heap available. JVMs always have an upper limit for memory (`Runtime.getRuntime().maxMemory()`), you cannot `new` things more than that. All that virtual memory and paging only applies to native allocations. If anyone, lib or app, does a `new` during runtime that'll chip away from the `Xmx`.

---
> The 64MB default was intentionally set high to cause ClassGraph to avoid writing to disk wherever possible.

I wholeheartedly support this, but for a 1k nested JAR there is no need to allocate and consume 64MB of Java heap. Which you fixed in 037983e. 🎉

---
> However if you were running into this case, were the inner jars in question deflated, not stored?

Yep, I checked inside `neo4j-lucene-upgrade-3.4.9.jar`, it has 3 jar files inside its `lib` folder. ZIP-ing a ZIP always gives a bit of extra gain. In this case it's 9% (500k) deflated vs. stored version. In principle I agree, JARs should always store inner JARs to be able to use optimizations for classpath like you did.

---
> re 0e944e0

Note: This alone wouldn't have helped, because I'm not using this library, Neo4J OGM does and I don't have access to `ClassGraph` builder. Luckily using the hint will help a lot in this case.

---
> I just checked in a large change that allows files to be backed with RandomAccessFile rather than MappedByteBuffer, eliminating the virtual memory pressure altogether. [...] so maybe I should remove the memory mapping altogether.

Memory mapping sounds like the optimal way to go, as it's handled natively, doesn't copy data (saves JVM heap and native memory). I guess in the end memory mapping still has to read from disk when the memory is accessed and because I/O is magnitude slower than RAM you won't see a difference in processing speed. The win is the change in amount and location of allocated memory.

**Comment by TWiStErRob** [comment]
I played around a bit more, I was curious if I can reproduce what I jokingly referred to

> So if all those are processing nested JARs, 2.5G of memory is required.

I changed the terminating code to:
```java
long used = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
System.out.println("Memory used: " + used / (1024 * 1024) + "M");
```
and gave the process much more memory with -Xmx10G

_The numbers are: before GC/after GC 4.8.62 &rarr; before GC/after GC 4.8.63-SNAPSHOT, in megabytes._
Without any dependency: 9M/2M &rarr; 9M/2M
With `runtimeOnly "org.neo4j:neo4j-lucene-upgrade:3.4.9"`: 208M/202M &rarr; 15M/4M
With `runtimeOnly files("libs/nested-stored.jar")`: 7M/3M &rarr; 7M/3M
With `runtimeOnly files("libs/nested-deflated.jar")`: 8482M/8453M &rarr; 14M/4M

The last two are artificial files with 128 nested JAR files, that I can't attach here :( I know it's unrealistic, but it proves your fix working, and it is working really well! See also PR #401.

**Comment by lukehutch** [comment]
> Wow, lot of info :) And also I'm not used to getting fixes (let alone replies) this fast. Well done!

You're welcome! That's the goal. Especially when somebody files a bug that is as detailed and deeply researched as yours is!

I usually try to ensure any bug that is reported is fixed within 24 hours, and a new release pushed out -- but this particular bug report has triggered a whole set of domino changes that affect ClassGraph at a very deep level, and it might take a few days to get all the changes in place. I'll try to get some more of that done tomorrow.

> > I hadn't considered though that Java limits not just the resident size but also the entire accessible virtual memory size
> 
> Yes, Java OOM means there's no more JVM heap available. JVMs always have an upper limit for memory (`Runtime.getRuntime().maxMemory()`), you cannot `new` things more than that. All that virtual memory and paging only applies to native allocations. If anyone, lib or app, does a `new` during runtime that'll chip away from the `Xmx`.

Right, but I thought you were saying `-Xmx` sets the maximum limit for all allocations, both heap and memory mapped files? Or do memory mappings fall outside the `-Xmx` limit (while still being limited, probably relative to RAM size)? (I guess I could just go test this, but it seems like you know the answer..)

> I wholeheartedly support this, but for a 1k nested JAR there is no need to allocate and consume 64MB of Java heap. Which you fixed in [037983e](https://github.com/classgraph/classgraph/commit/037983ee0a876a77b0b26050a00558d35d965d90). 

Yes, this was a major oversight, that I forgot to trim that array! I think that alone is the source of the sometimes dramatic drop in memory consumption you saw in your latest comment.

> Yep, I checked inside `neo4j-lucene-upgrade-3.4.9.jar`, it has 3 jar files inside its `lib` folder. ZIP-ing a ZIP always gives a bit of extra gain. In this case it's 9% (500k) deflated vs. stored version. In principle I agree, JARs should always store inner JARs to be able to use optimizations for classpath like you did.

I assumed that zipping a zipfile rarely if ever resulted in further compression, but I guess if a zipfile already contained stored uncompressed content, it should be compressible. Or maybe some zipfiles contain central directories that are large enough that zip thinks it is worth it to zip the file for the compression gains. However I just looked at my `~/.m2/repository` cache, and there isn't a single jar in there that contains a nested jar. So I think it is only web applications, Spring-Boot applications, etc. (with custom classloaders) that cause this issue to arise, and not libraries in general.
 
> Note: This alone wouldn't have helped, because I'm not using this library, Neo4J OGM does and I don't have access to `ClassGraph` builder. Luckily using the hint will help a lot in this case.

Well that's even more amazing then that you put so much time and effort into tracking down the source of the problem, when it was a dependency of a dependency that was causing the problem!

> Memory mapping sounds like the optimal way to go, as it's handled natively, doesn't copy data (saves JVM heap and native memory).

I always thought so before, but I'm not so sure anymore. Actually the new `RandomAccessFile` method only ever relies on operating system buffers, which will only ever use unused RAM, so the memory overhead is basically zero.

> I guess in the end memory mapping still has to read from disk when the memory is accessed and because I/O is magnitude slower than RAM you won't see a difference in processing speed. The win is the change in amount and location of allocated memory.

Yes, and this is done lazily. But I didn't know that OS file buffering worked so well with `RandomAccessFile` -- I thought it would cause many additional disk reads, leading to greater overhead.

I should test this on at least Windows before I totally pull the plug on `MappedByteBuffer`, but to be honest the mmap support in ClassGraph has caused a neverending series of headaches over the years, e.g. because without using `misc.Unsafe`, you can't forcibly unmap mapped byte buffers, you just have to wait for the GC to decide to collect them, then they're freed -- and if you don't unmap them, Windows holds a file lock on them, which can cause files that were supposed to be deleted to be left behind. Anyway although ClassGraph seems to work flawlessly with mmap now across all three major operating systems (and it took a lot of work to get there), I won't shed any tears ripping all of it out as long as the performance is not impacted.

> The last two are artificial files with 128 nested JAR files, that I can't attach here :( I know it's unrealistic, but it proves your fix working, and it is working really well! See also PR #401.

Thanks, I really appreciate the contribution!

**Comment by lukehutch** [comment]
PS can you please test the latest SNAPSHOT version's use of `RandomAccessFile` with the following?

```
new ClassGraph().disableMemoryMapping()
```

Performance and results should not be impacted, and memory usage should be the same as (or better than) the results in your previous comment.

**Comment by lukehutch** [comment]
Hi, I just checked in a huge change that completely reaches the virtual filesystem layer to remove mmap support. This should give you much lower memory overhead, with maybe up to a 10% decrease in performance. Can you please check if the git master version works for you?

**Comment by lukehutch** [comment]
@TWiStErRob The fixes are released as [4.8.63](https://github.com/classgraph/classgraph/releases/tag/classgraph-4.8.63). You should find that the memory pressure is substantially reduced. Please let me know if you have any issues with this.

I really appreciate your bug report, the extensive testing, the unit test, the comments, and the testing -- thank you!

If this works for you, you should convince the Neo4J-OGM people to bump their minimum version requirement.  /cc: @michael-simons

**Comment by michael-simons** [comment]
Yep. Open a ticket with us please, @TWiStErRob

**Comment by TWiStErRob** [comment]
Hi, sorry I haven't found the time to try it before, but I would've done the same thing as #401 does anyway. There's an issue with the latest release, due to Java binary incompatibility.

I've replaced the dependency in my project to try it out:
```gradle
allprojects {
	configurations.all {
		resolutionStrategy.dependencySubstitution.all {
			val requested = this.requested
			if (requested is ModuleComponentSelector
				&& requested.module == "classgraph"
				&& requested.group == "io.github.classgraph"
			) {
				if (VersionNumber.parse(requested.version) < VersionNumber.parse("4.8.62")) {
					useTarget(
						"${requested.group}:${requested.module}:4.8.63",
						"Performance issue with low memory: https://github.com/classgraph/classgraph/issues/400"
					)
				}
			}
		}
	}
}
```
I'll report a new issue for this.

**Comment by TWiStErRob** [comment]
After working around #403 by using JDK 9, I can confirm that a complex Neo4J OGM, Kotlin, Ktor app runs with Xmx64M happily, and Xmx32M still runs into OOM, but doesn't spin.

**Comment by lukehutch** [comment]
Thanks for testing, @TWiStErRob! #403 is fixed and released in 4.8.64.

**Comment by TWiStErRob** [comment]
Raised https://github.com/neo4j/neo4j-ogm/pull/762

**Comment by michael-simons** [comment]
Fantastic thread. 

FWIW: You having `org.neo4j:neo4j-lucene-upgrade:3.4.9` on the class path, which probably means you're using Neo4j-OGM embedded. Please consider if this is really necessary.
We won't support 4.0 in Neo4j-OGM embedded.

If it is only for testing, I recommend sticking with Neo4j-OGM Bolt transport and configure an embedded instance for testing separately (by using our Test Harness). 

Here are two examples in case of a Spring Boot application:
https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/tree/master/examples/ogm-integration/src/test/java/org/neo4j/doc/driver/springframework/boot/ogm_integration/domain

If you don't go with Test containers, consider Test harness, something like this: https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/blob/master/examples/ogm-integration/src/test/java/org/neo4j/doc/driver/springframework/boot/ogm_integration/domain/TestSliceTestWithHarness.java#L54

Would work in Ktor as well.

The benefit of it being that you decouple the embedded database from your object mapping framework.

**Comment by TWiStErRob** [comment]
Huh, thanks @michael-simons, I think I'm using harness now; but the embedded dependency may still be there as I was migrating an older app. I'll double-check my classpath.

**Comment by michael-simons** [comment]
Harness has of course the whole Neo4j dependency graph anyway. 

My main points:

* Don't couple embedded creation to OGM 
* Don't run Bolt in prod and Neo4j-OGM embedded in test… 

Just bring up Harness and configure OGM to use the Bolt port of Harness.

Thanks :)
